### PR TITLE
#306 prepare back end build

### PR DESCRIPTION
**Pull request by andreievg**
Fixes: #306 

There is quite a bit of diff, but the bulk of it is in plugin changes (which are quite minor and repetitive).
I've tried to minimise the diff with `git mv`, but still some file were not recognised as moves and as deletes : /

### Changes

* Relocated plugins and images to root
* Remove dependencies from plugins, remove jest and tsconfig from plugins
* Package json in each plugin to use scripts in `utils/pluginScripts` (using root jest and tsc at root level)
* Relocate plugin.json to plugin root
* Remove reliance on function name for plugins (each plugin should have 'action' exported via index.js)
* Added post install script to auto yarn install all plugins when global yarn install is called
* Updated docs
* Global yarn build will build everything and move built plugins to overall build folder
* Updates to use new folder structure, including using `getAppEntryPointDir()` vs `getAppRootDir()`, maybe I could have used existing entry point function, but had some issues with it in built environment, happy to change back if there is a work around. (not super happy with the way I've changed it tbh, but it works)
 
##### utils/pluginScripts
* `buildPlugin.js` -> installs and build a plugin, saving output to `build/{pluginName}` (within plugin folder)
  * adds temporary tsconfig.json to build just the plugin
  * builds in temporary folder and relocates build content to `build/{pluginName}` (within plugin folder)
  * cleans up
* `buildPluginFromPluginFolder.js` -> checks working directory to identify plugin name and uses `buildPlugins.js` to build plugin
* `testPluginFromPluginFolder.js` -> checks working directory to identify plugin name and runs `yarn test {pluginname}`
* `yarnInstallPlugins.js` -> runs `yarn install` with working directory being plugin folder (for all plugins asynchronously), used in `postinstall` of global package.json

### Testing

I did basic test with both build version (after `yarn build` -> `node build/src/server.js`) and normal dev version.

Was watching console output for problems and went through making a few applications (making sure to check at least one file upload, making sure thumbnail shows correctly)

Also made sure that correct console log is shown for trim responses when application is submitted (to make sure not errors are thrown for deep-equal package, as it's dependency specific to that plugin)

### CODE REVIEW

**Comment by CarlosNZ** [comment]
One quick comment: most of the plugin folders have all their dependencies in their folder root. A couple of them (e.g. update_review_assignments) have a `node_modules` folder, which seems much tidier.

But shouldn't `node_modules` be in .gitignore anyway? (Maybe each plugin needs its own .gitignore file?)

File: documentation/Action-plugin-specification.md [hunk 1]
```diff
@@ -8,26 +8,28 @@ See [here](./List-of-Action-plugins.md) for a list of currently implemented Acti
...
-- `/src/<plugin-name>.ts` -- Typescript source code
-- `/src/<plugin-name>.js` -- Javascript code compiled from Typescript. This is what is actually loaded and run in the built server app.
+- `plugin.json` -- contains metadata for the plugin. The system reads this file (at startup) to load all the relevant information about the plugin into the database, so make sure it is accurate and up-to-date. Further details [below](#plugin).
+- `/src/index.ts` -- A file that export the plugin entrypoint method as `action`
+- `/src/*.ts` -- All other plugin files
 
 ## Source code file
 
-The main source code for the plugin is simply an exported function that the server app dynamically loads via `require` at run-time. Calling the function exported by each action plugin, sequentially, is the job of the **Action module**.
+Plugin can have any number of source files located in the source folder, there must be an exported entrypoint method in `index.ts`, this method will be dynamically loaded via `require` at run-time
```
Comment (CarlosNZ) [inline documentation/Action-plugin-specification.md:23]:
```suggestion
Plugin can have any number of source files located in the source folder, but there must be an exported entrypoint method in `index.ts`. This method will be dynamically loaded via `require` at run-time.
```

File: documentation/Action-plugin-specification.md [hunk 2]
```diff
@@ -8,26 +8,28 @@ See [here](./List-of-Action-plugins.md) for a list of currently implemented Acti
...
+Calling the function exported by each action plugin, sequentially, is the job of the **Action module**.
 
-Plugin code should follow this basic structure:
+Plugin entrypoint method should follow this basic structure:
+
+TO_DO describe `parameters`
```
Comment (CarlosNZ) [inline documentation/Action-plugin-specification.md:28]: Yeah, I'll do this as part of #282

File: documentation/Action-plugin-specification.md [hunk 3]
```diff
@@ -42,6 +44,8 @@ module.exports['consoleLog'] = function (parameters: any) {
     }
   }
 }
+// src/index.ts
+export {default as action} from './consoleLog'
```
Comment (CarlosNZ) [inline documentation/Action-plugin-specification.md:48]: We should probably re-write this more like a "template", with just some generic placeholder inside the `try`method, etc. I'll do this with #282.

File: documentation/Action-plugin-specification.md [hunk 4]
```diff
@@ -110,28 +110,42 @@ The `action_plugin` table is the primary record of what actions are available to
...
-- `function_name`
+- `path` (determined by the folder the plugin was found in)
 - `required_parameters`
 
 <a name="development"></a>
 
 ## Development and build process
 
-Plugins are intended to be stand-alone units, in that a user should be able to simply copy a plugin folder to another system and have it work straight away. (Eventually, we envisage that plugins could be imported directly via the front-end UI.) To that end, each plugin is developed as its own package, with its own `package.json` and `node_modules` folder, as well as Typescript and testing configurations.
+Plugins are intended to be stand-alone units, in that a user should be able to simply `yarn build` and copy content inside `build` folder to another system and have it work straight away. (Eventually, we envisage that plugins could be imported directly via the front-end UI.) To that end, each plugin is developed as its own package, with its own `package.json` and `node_modules` folder. Root typescript and testing configurations are used by default but can be overwritten
```
Comment (CarlosNZ) [inline documentation/Action-plugin-specification.md:120]: It doesn't seem to work like this. I just tried running `yarn build` inside a plugin folder and it failed. Also -- when I just run `yarn install` within plugin folder, it says "already up-to-date", but there's no node_modules folder in there.

File: package.json [hunk 5]
```diff
@@ -33,17 +33,18 @@
...
-    "build": "rimraf ./build && tsc && rsync -a --exclude \"*/node_modules\" ./src/plugins ./build",
+    "build": "./utils/build_all.sh",
     "test": "jest --runInBand",
     "pg": "exec ./utils/postgraphile_launch.sh",
     "pg_permissions": "exec ./utils/postgraphile_launch_permissions.sh",
     "dev_pg": "concurrently --kill-others-on-fail \"yarn dev\"  \"yarn pg\"",
     "dev_pg_permissions": "concurrently --kill-others-on-fail \"yarn dev\"  \"yarn pg_permissions\"",
-    "database_init": "./database/initialise_database.sh && ./database/insert_data.sh && ./database/post_data_insert.sh",
-    "build_plugins": "exec ./utils/build_plugins.sh",
+    "database_init": "exec ./database/initialise_database.sh",
```
Comment (CarlosNZ) [inline package.json:42]: Was this an intentional re-version of the database_init script? Don't you need that for the snapshots?

File: package.json [hunk 6]
```diff
@@ -33,17 +33,18 @@
...
+    "build_plugins": "exec ./utils/pluginScripts/build_plugins.sh",
```
Comment (CarlosNZ) [inline package.json:43]: This file doesn't exist. Did you perhaps mean `build_all.sh`?

File: documentation/Action-plugin-specification.md (hunk 4 above)
Comment (CarlosNZ) [reply documentation/Action-plugin-specification.md:120]: Okay, it's working now (had to yarn install at top level). But build all still not working (see comment below -- wrong filename I think)

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
Looking good, documentation is helpful, thanks. Just a couple of little comments in the code

Tried `yarn build` on the whole app and it seems to work well.

One thing not working properly for me is the tests. Tried to run `yarn test` from within a plugin folder and I get:
```
 FAIL  plugins/action_trim_responses/build/action_trim_responses/src/trimResponses.test.js
  ● Test suite failed to run
```

One suggestion: you've got a "build_plugins" script for building just the plugins. Wonder if we need an additional flag on the main "build" script to tell it not to re-build plugins if they're already built -- would save quite a bit of time in the build. Maybe `yarn build -noplugins` or something? I dunno, probably not a high priority for now.

File: package.json (hunk 5 above)
Comment (andreievg) [reply package.json]: That's my clumsy conflict resolution during merge from master:
48190f2a4a08ffcf254b7b08afec66c6f4297c4a, thanks for noticing

File: package.json (hunk 6 above)
Comment (andreievg) [reply package.json]: oops, thanks for noticing:

396a243682d04d1cef5d6209c85397ad70ed808e

File: documentation/Action-plugin-specification.md (hunk 4 above)
Comment (andreievg) [reply documentation/Action-plugin-specification.md:120]: > It doesn't seem to work like this. I just tried running yarn build inside a plugin folder and it failed. Also -- when I just run yarn install within plugin folder, it says "already up-to-date", but there's no node_modules folder in there.

Hmm, if root level node_modules are inline with root level package.json, should be able to `yarn build` from within plugin folder, I've just checked and it seems to work.

Comment (CarlosNZ) [reply documentation/Action-plugin-specification.md:120]: Yeah, it works now (see prev comment).

**Comment by andreievg** [comment]
> One thing not working properly for me is the tests. Tried to run yarn test from within a plugin folder and I get:

I think i stuffed up the tests completely. Will fix shortly

**Comment by andreievg** [comment]
> One thing not working properly for me is the tests. Tried to run yarn test from within a plugin folder and I get:

I did some fixes:

* make sure serial number update is only done for snapshots
* change tests to use correct import
* updated ID on one of the tests (createUser)
* jest config to ignore build folder (we would need another jest config for build testing)

It looks like there are quite a lot of id mismatches now (trim action test cant' even insert data for the test), I think we need to make test more resilient to data changes, I've suggested something here: https://github.com/openmsupply/application-manager-server/blob/master/documentation/Snapshots.md#todo, I'll make and issue over the weekend, but the idea is to load snapshot (which is stored for each test or globally) before running test

**Comment by andreievg** [comment]
54dd302210c522438cfc56a4755908b2ed5dda05

**Comment by CarlosNZ** [comment]
Cool, tests are working now (by which I mean they run -- not necessarily pass lol).

**Review by CarlosNZ — COMMENTED** [review]
Cool, looking good now. My only concern is reconciling this with the few small changes to the action plugins that we've done over the last few days that won't be in here. How should we check?

**Comment by andreievg** [comment]
>Cool, looking good now. My only concern is reconciling this with the few small changes to the action plugins that we've done over the last few days that won't be in here. How should we check?

In github desktop, I selected this branch, then went to 'commits', found the first commit of this branch. used commit just before this branch started and compared master with it: 

https://github.com/openmsupply/application-manager-server/compare/bc3d41215a80bff953fa2e027bdc99f299b7721f...master. Couldn't see any src/plugins in the file changed list

**Comment by CarlosNZ** [comment]
> [bc3d412...master](https://github.com/openmsupply/application-manager-server/compare/bc3d41215a80bff953fa2e027bdc99f299b7721f...master). Couldn't see any src/plugins in the file changed list

There must be though, cos at the very least you've changed the "wrapper" for each and the export setting.

**Comment by andreievg** [comment]
> There must be though, cos at the very least you've changed the "wrapper" for each and the export setting.

I was comparing master with 'master' but before this branch was started, to see if we've done any changes to the plugins, doesn't look like it

**Comment by CarlosNZ** [comment]
Ah, okay cool

**Review by CarlosNZ — APPROVED** [review]

**Pull request merged**